File: runtime/ruby/units/wasi/pack_filestat.rb
# requires: wasi/wasi_filetype
# Packs a File::Stat into a WASI filestat (64 bytes): dev, ino, filetype
# (+7 pad), nlink, size, atim/mtim/ctim (all u64, times in nanoseconds).
def pack_filestat(stat)
  [
    stat.dev, stat.ino, wasi_filetype(stat), stat.nlink, stat.size,
    (stat.atime.to_r * 1_000_000_000).to_i,
    (stat.mtime.to_r * 1_000_000_000).to_i,
    (stat.ctime.to_r * 1_000_000_000).to_i
  ].pack("Q<Q<Cx7Q<Q<Q<Q<Q<")
end
private :pack_filestat

File: runtime/perl/units/wasi/pack_filestat.pl
# requires: wasi/wasi_filetype
# Packs a Time::HiRes::stat result (arrayref) into a WASI filestat (64
# bytes): dev, ino, filetype (+7 pad), nlink, size, atim/mtim/ctim (all
# u64, times in nanoseconds — quantized to what NV seconds can carry,
# roughly 400ns granularity at the current epoch).
sub pack_filestat {
    my ($self, $st) = @_;
    return pack('Q<Q<Cx7Q<Q<Q<Q<Q<',
        $st->[0], $st->[1], $self->wasi_filetype($st->[2]), $st->[3], $st->[7],
        map { int($_ * 1e9 + 0.5) } @{$st}[8, 9, 10]);
}
